### PR TITLE
Add wait group to ratelimit operator

### PR DESCRIPTION
**Pull request by djaglowski**
## **Please check that the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Add a changelog entry (for non-trivial bug fixes / features)
- [ ] CI passes

### CODE REVIEW

**Comment by djaglowski** [comment]
| Log Files | Logs / Second | CPU Avg (%) | CPU Avg Δ (%) | Memory Avg (MB) | Memory Avg Δ (MB) |
| --- | --- | --- | --- | --- | --- |
|1|1000|1.6207083|-0.10345805|127.257675|-2.3106232|
|1|5000|5.9311566|-0.15516376|135.96498|+0.8643646|
|1|10000|12.29332|+0.2238636|144.09483|-0.5001221|
|1|50000|56.932117|-3.273674|172.88039|+1.5065918|
|1|100000|119.122215|+5.3502426|229.18346|-13.894669|
|10|100|2.4483037|+0.22398496|132.73666|+2.1919403|
|10|500|6.7242556|-0.37930775|137.02129|+3.5133362|
|10|1000|13.379681|-0.9311428|145.38739|+2.381195|
|10|5000|62.96997|-1.1866379|182.10332|-2.1749725|
|10|10000|124.060905|-0.64863586|226.95663|+0.5991516|

File: operator/builtin/transformer/ratelimit/rate_limit.go [hunk 1]
```diff
@@ -66,6 +67,7 @@ type RateLimitOperator struct {
 	burst    uint
 	isReady  chan struct{}
 	cancel   context.CancelFunc
+	wg       *sync.WaitGroup
```
Comment (camdencheek) [inline operator/builtin/transformer/ratelimit/rate_limit.go:70]: No need for a pointer here

**Review by camdencheek — COMMENTED** [review]
Looks to me like an additional issue here is that we can hang on line 98 trying to add a to `p.isReady` if `p.isReady` is full, so that goroutine may never exit, causing `Stop()` to never return.

File: operator/builtin/transformer/ratelimit/rate_limit.go [hunk 2]
```diff
@@ -84,7 +84,7 @@ func (p *RateLimitOperator) Start() error {
 
 	ctx, cancel := context.WithCancel(context.Background())
 	p.cancel = cancel
-	p.wg = &sync.WaitGroup{}
+	p.wg = sync.WaitGroup{}
```
Comment (camdencheek) [inline operator/builtin/transformer/ratelimit/rate_limit.go:87]: Now that this is not a pointer, no need to initialize it -- the zero value of `sync.WaitGroup` is valid

**Comment by djaglowski** [comment]
| Log Files | Logs / Second | CPU Avg (%) | CPU Avg Δ (%) | Memory Avg (MB) | Memory Avg Δ (MB) |
| --- | --- | --- | --- | --- | --- |
|1|1000|1.7069275|+0.034486175|127.24394|-1.7861023|
|1|5000|6.172591|+0.1897564|134.59254|-1.3744507|
|1|10000|12.224411|-0.10360432|141.93008|+0.014266968|
|1|50000|63.553402|+3.775402|176.9472|+0.4074707|
|1|100000|120.5168|+5.9316864|223.88995|-9.034882|
|10|100|2.2586448|+0.03447032|133.217|+0.44099426|
|10|500|7.017362|+0.53451395|142.1774|+4.8354034|
|10|1000|13.552072|-0.069340706|143.85413|+1.2118835|
|10|5000|63.56547|-0.06404114|180.0493|-5.6749725|
|10|10000|125.25264|-3.2819824|229.93965|-3.7528381|

**Comment by djaglowski** [comment]
| Log Files | Logs / Second | CPU Avg (%) | CPU Avg Δ (%) | Memory Avg (MB) | Memory Avg Δ (MB) |
| --- | --- | --- | --- | --- | --- |
|1|1000|1.6724454|+4.053116e-06|129.21808|+0.18803406|
|1|5000|6.1207967|+0.13796186|136.83621|+0.8692169|
|1|10000|12.155769|-0.17224598|145.25781|+3.3419952|
|1|50000|59.570274|-0.20772552|176.57637|+0.036636353|
|1|100000|113.29298|-1.2921371|233.70259|+0.77775574|
|10|100|2.379409|+0.15523458|132.32947|-0.4465332|
|10|500|6.810442|+0.3275938|136.47656|-0.86543274|
|10|1000|13.310617|-0.31079483|140.7073|-1.9349365|
|10|5000|63.457695|-0.17181778|182.89548|-2.8287964|
|10|10000|122.69238|-5.8422394|223.36247|-10.330017|

**Pull request merged**